src/Bicep.Core/Semantics/DeclarationVisitor.cs

```diff
@@ -36,65 +36,91 @@ public static (ImmutableArray<DeclaredSymbol>, ImmutableArray<LocalScope>) GetAl
             return (declarations.ToImmutableArray(), childScopes.Select(MakeImmutable).ToImmutableArray());
         }
 
+        public override void VisitProgramSyntax(ProgramSyntax syntax)
+        {
+            base.VisitProgramSyntax(syntax);
+        }
+
         public override void VisitParameterDeclarationSyntax(ParameterDeclarationSyntax syntax)
         {
             base.VisitParameterDeclarationSyntax(syntax);
 
             var symbol = new ParameterSymbol(this.context, syntax.Name.IdentifierName, syntax, syntax.Modifier);
-            this.declaredSymbols.Add(symbol);
+            DeclareSymbol(symbol);
         }
 
         public override void VisitVariableDeclarationSyntax(VariableDeclarationSyntax syntax)
         {
             base.VisitVariableDeclarationSyntax(syntax);
 
             var symbol = new VariableSymbol(this.context, syntax.Name.IdentifierName, syntax, syntax.Value);
-            this.declaredSymbols.Add(symbol);
+            DeclareSymbol(symbol);
         }
 
         public override void VisitResourceDeclarationSyntax(ResourceDeclarationSyntax syntax)
         {
+            // Create a scope for each resource - this ensures that nested resources 
+            // are contained within the appropriate scope.
+            var scope = new LocalScope(string.Empty, syntax, ImmutableArray<DeclaredSymbol>.Empty, ImmutableArray<LocalScope>.Empty);
```

Comment:
Note: I've made no attempt to *fuse* this with the scope created for a loop. So if you have a loop, I think this will result in two scopes being created. Need to test this an make sure the loop's scope doesn't hide the resource's symbol

Reply:
Btw, I made some changes to binding of locals in my loop branch to allow something like this `[for thing in thing: thing]`. In that branch there's is a separation between the syntax node that declares the scope from the node that actually affects binding. In the case of loops `ForSyntax` declares the locals, but its Value (`ObjectSyntax` for resources) is what affects binding. This effectively means that the loop array expression binds against the parent scopes but `VariableAccessSyntax` can bind to the loop variable inside the loop body.

If we don't attempt to fuse the scopes, one of the identifiers will be hidden in the child scopes.

---
In reply to: [577218237](https://github.com/Azure/bicep/pull/1516#discussion_r577218237) [](ancestors = 577218237)

Reply:
> If we don't attempt to fuse the scopes, one of the identifiers will be hidden in the child scopes.
>
> In that branch there's is a separation between the syntax node that declares the scope from the node that actually affects binding.

Here's what we might want to do then, we might want to associate the scope declaration with the ResourceDeclarationSyntax (my change does this), but have it take effect within the `ObjectSyntax` (not included for my PR, but in yours).

Then when you're processing a loop for a resource, you can lookup the scope that is already *bound* to it, and party on that. That seems like a simple protocol for these two changes to work together. Since my change is operating a higher level than yours in the processing, it makes sense that it would happen in that order.

Reply:
I'm already using the `ObjectSyntax` under a `ForSyntax` as the "binding" node btw :)

Let's consider this (duplicate identifiers are intentional 😈):
```bicep
var thing = []
resource parent '...' = [for thing in thing: {
  resource thing '...' = {
    ...
  }
}]
```

Are you proposing the following hierarchy in the AST for the top-level resource (going outside in)?
- `ResourceDeclarationSyntax` (resource scope declaring node)
- `ForSyntax` (resource scope binding node; loop scope declaring node)
- `ObjectSyntax` (loop node binding node)
- `ResourceDeclarationSyntax`
- `ObjectSyntax`
- ...

If yes, then I'm with you. Going through this made me realize why we shouldn't try to fuse the scopes because then we end up with an argument about who declared it (or have two declaring nodes for a local scope, which is even weirder).

Reply:
This is what I ended up going with. The Resource Declaration will bind its scope to `syntax.Value`.